packages/core/contracts/oracle/implementation/Proposer.sol

```diff
@@ -0,0 +1,94 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+pragma solidity ^0.8.0;
+
+import "./Finder.sol";
+import "./Governor.sol";
+import "./Constants.sol";
+import "./Voting.sol";
+import "./AdminIdentifierLib.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+/**
+ * @title Proposer contract that allows anyone to make governance proposals with a bond.
+ */
+contract Proposer is Ownable, Testable {
+    using SafeERC20 for IERC20;
+    IERC20 public token;
+    uint256 public bond;
+    Governor public governor;
+    Finder public finder;
+
+    struct BondedProposal {
+        address sender;
+        // 64 bits to save a storage slot.
+        uint64 time;
+        uint256 lockedBond;
+    }
+    mapping(uint256 => BondedProposal) public bondedProposals;
+
+    event BondSet(uint256 bond);
+    event ProposalResolved(uint256 indexed id, bool success);
+
+    /**
+     * @notice Construct the Proposer contract.
+     */
+    constructor(
+        IERC20 _token,
+        uint256 _bond,
+        Governor _governor,
+        Finder _finder,
+        address _timer
+    ) Testable(_timer) {
+        token = _token;
+        bond = _bond;
+        governor = _governor;
+        finder = _finder;
+        emit BondSet(_bond);
+    }
+
+    /**
+     * @notice Propose a new set of governance transactions for vote.
+     * @dev Pulls bond from the caller.
+     * @param transactions list of transactions for the governor to execute.
+     * @return id the id of the governor proposal.
+     */
+    function propose(Governor.Transaction[] memory transactions) public returns (uint256 id) {
```

Comment:
Should we reentrancy-guard these public methods?

Reply:
Probably!

Reply:
Done.